File: src/codexcode/accept.py
"""Applying a result back to the original working tree.

We support three apply modes:

  - `--from host`         copy every host-touched file into the original tree
  - `--from challenger`   same for the challenger
  - `--files SPEC,SPEC..` per-file selection, e.g. `host:src/a.py,challenger:src/b.py`

Apply only mutates the working tree; the host agent is expected to commit
afterwards (or invoke `codexcode commit`). Cleanup is a separate step so the
caller can inspect or amend before tearing down.
"""
from __future__ import annotations

import os
import pathlib
import shutil
from typing import Dict, List, Tuple

from . import state
from .isolation import changed_files, remove_worktree


def _side_worktree(meta: dict, side: str) -> str:
    if side == state.HOST:
        return meta["host"]["worktree"]
    if side == state.CHALLENGER:
        return meta["challenger"]["worktree"]
    raise ValueError(f"unknown side: {side}")


def _all_changed(meta: dict, side: str) -> List[Tuple[str, str]]:
    return changed_files(_side_worktree(meta, side), meta["base_commit"])


def _copy_file(src_root: str, dst_root: str, rel_path: str) -> None:
    src = pathlib.Path(src_root) / rel_path
    dst = pathlib.Path(dst_root) / rel_path
    dst.parent.mkdir(parents=True, exist_ok=True)
    if src.exists():
        shutil.copy2(src, dst)
    else:
        if dst.exists():
            dst.unlink()


def _delete_file(dst_root: str, rel_path: str) -> None:
    dst = pathlib.Path(dst_root) / rel_path
    if dst.exists():
        dst.unlink()


def apply_side(meta: dict, side: str) -> Dict[str, List[str]]:
    repo = meta["repo_root"]
    wt = _side_worktree(meta, side)
    touched = _all_changed(meta, side)
    applied = []
    deleted = []
    for status, rel_path in touched:
        if status.startswith("D"):
            _delete_file(repo, rel_path)
            deleted.append(rel_path)
        else:
            _copy_file(wt, repo, rel_path)
            applied.append(rel_path)
    return {"applied": applied, "deleted": deleted}


def parse_files_spec(spec: str) -> List[Tuple[str, str]]:
    """`host:a.py,challenger:b.py` -> [('host','a.py'), ('challenger','b.py')]"""
    out: List[Tuple[str, str]] = []
    for chunk in spec.split(","):
        chunk = chunk.strip()
        if not chunk:
            continue
        if ":" not in chunk:
            raise ValueError(
                f"bad --files entry {chunk!r}; expected SIDE:PATH"
            )
        side, _, path = chunk.partition(":")
        side = side.strip().lower()
        path = path.strip()
        if side not in (state.HOST, state.CHALLENGER):
            raise ValueError(
                f"unknown side {side!r} in {chunk!r}; want host or challenger"
            )
        if not path:
            raise ValueError(f"empty path in {chunk!r}")
        out.append((side, path))
    return out


def apply_files(
    meta: dict, picks: List[Tuple[str, str]]
) -> Dict[str, List[str]]:
    repo = meta["repo_root"]
    applied: List[str] = []
    deleted: List[str] = []
    for side, rel_path in picks:
        wt = _side_worktree(meta, side)
        src = pathlib.Path(wt) / rel_path
        if src.exists():
            _copy_file(wt, repo, rel_path)
            applied.append(f"{side}:{rel_path}")
        else:
            _delete_file(repo, rel_path)
            deleted.append(f"{side}:{rel_path}")
    return {"applied": applied, "deleted": deleted}


def cleanup_session(session_id: str) -> Dict[str, List[str]]:
    """Remove both worktrees, both ephemeral branches, and the session dir."""
    meta = state.load_meta(session_id)
    repo = meta["repo_root"]
    removed: List[str] = []
    for side in (state.HOST, state.CHALLENGER):
        side_meta = meta[side if side == state.HOST else "challenger"]
        wt = side_meta["worktree"]
        branch = side_meta["branch"]
        remove_worktree(repo, wt, branch)
        removed.append(wt)
    state.remove_session(session_id)
    return {"removed": removed}

File: src/accept.ts
import {
  copyFileSync,
  existsSync,
  mkdirSync,
  rmSync,
  statSync,
  unlinkSync,
  chmodSync,
} from "node:fs";
import { dirname, join, posix } from "node:path";
import { CHALLENGER, HOST, type Side } from "./constants.js";
import { changedFiles, removeWorktree } from "./isolation.js";
import { removeSession, type SessionMeta } from "./state.js";

function sideWorktree(meta: SessionMeta, side: Side): string {
  return side === HOST ? meta.host.worktree : meta.challenger.worktree;
}

export function validateRelativePath(relPath: string): string {
  const normalized = relPath.replace(/\\/g, "/");
  const parsed = posix.normalize(normalized);
  if (
    !normalized ||
    posix.isAbsolute(normalized) ||
    parsed === "." ||
    parsed.startsWith("../") ||
    parsed === ".." ||
    normalized.split("/").some((part) => part === "" || part === "." || part === "..")
  ) {
    throw new Error(`path must be a relative path inside the repo: ${JSON.stringify(relPath)}`);
  }
  return parsed;
}

function copyFile(srcRoot: string, dstRoot: string, relPath: string): void {
  const safePath = validateRelativePath(relPath);
  const src = join(srcRoot, safePath);
  const dst = join(dstRoot, safePath);
  mkdirSync(dirname(dst), { recursive: true });
  if (existsSync(src)) {
    copyFileSync(src, dst);
    chmodSync(dst, statSync(src).mode);
  } else if (existsSync(dst)) {
    unlinkSync(dst);
  }
}

function deleteFile(dstRoot: string, relPath: string): void {
  const safePath = validateRelativePath(relPath);
  const dst = join(dstRoot, safePath);
  if (existsSync(dst)) {
    unlinkSync(dst);
  }
}

export function applySide(meta: SessionMeta, side: Side): { applied: string[]; deleted: string[] } {
  const repo = meta.repo_root;
  const wt = sideWorktree(meta, side);
  const touched = changedFiles(wt, meta.base_commit);
  const applied: string[] = [];
  const deleted: string[] = [];
  for (const [status, relPath] of touched) {
    if (status.startsWith("D")) {
      deleteFile(repo, relPath);
      deleted.push(relPath);
    } else {
      copyFile(wt, repo, relPath);
      applied.push(relPath);
    }
  }
  return { applied, deleted };
}

export function parseFilesSpec(spec: string): Array<[Side, string]> {
  const out: Array<[Side, string]> = [];
  for (const chunkRaw of spec.split(",")) {
    const chunk = chunkRaw.trim();
    if (!chunk) {
      continue;
    }
    if (!chunk.includes(":")) {
      throw new Error(`bad --files entry ${JSON.stringify(chunk)}; expected SIDE:PATH`);
    }
    const [sideRaw, ...pathParts] = chunk.split(":");
    const side = sideRaw.trim().toLowerCase();
    const relPath = pathParts.join(":").trim();
    if (side !== HOST && side !== CHALLENGER) {
      throw new Error(`unknown side ${JSON.stringify(side)} in ${JSON.stringify(chunk)}; want host or challenger`);
    }
    if (!relPath) {
      throw new Error(`empty path in ${JSON.stringify(chunk)}`);
    }
    out.push([side, validateRelativePath(relPath)]);
  }
  return out;
}

export function applyFiles(
  meta: SessionMeta,
  picks: Array<[Side, string]>,
): { applied: string[]; deleted: string[] } {
  const applied: string[] = [];
  const deleted: string[] = [];
  for (const [side, relPath] of picks) {
    const wt = sideWorktree(meta, side);
    const src = join(wt, relPath);
    if (existsSync(src)) {
      copyFile(wt, meta.repo_root, relPath);
      applied.push(`${side}:${relPath}`);
    } else {
      deleteFile(meta.repo_root, relPath);
      deleted.push(`${side}:${relPath}`);
    }
  }
  return { applied, deleted };
}

export function cleanupSession(sessionId: string, meta: SessionMeta): { removed: string[] } {
  const removed: string[] = [];
  for (const side of [HOST, CHALLENGER] as const) {
    const sideMeta = side === HOST ? meta.host : meta.challenger;
    removeWorktree(meta.repo_root, sideMeta.worktree, sideMeta.branch);
    rmSync(sideMeta.worktree, { recursive: true, force: true });
    removed.push(sideMeta.worktree);
  }
  removeSession(sessionId);
  return { removed };
}
